### ISSUE TITLE
complex dtype not supported

### ISSUE THREAD

**Issue by kharris** [classner/pymp#1]
This would be nice for my usage case

**Comment by classner** [comment]
Can you specify more exactly where you're missing this functionality? A few thoughts on it in general: the forking and processing will be really fast, thanks to the system copy-on-write strategy. The only but critical requirement for this is, that the loop variable objects are small, because they will be transferred between the threads to coordinate execution. Large objects can be created in memory either pre-forking if they're all the same between the threads, or post-forking if they're thread specific, and then there will be no overhead.

**Comment by kharris** [comment]
I'd like to be able to run something like
`
y = pymp.shared.array((m, n), dtype=np.complex)
with pymp.Parallel(self.num_threads) as p:
    for col in p.range(n):
        y[:,col]=fun(args)
`
where my shared memory array is full of complex numbers.

**Comment by kharris** [comment]
I was able to cheat it by breaking the array into real and imaginary parts, but it would be nice to have arbitrary numpy types supported:
```
xout_real = pymp.shared.array((self.number_of_nodes(), n_replica),  dtype='float64')
xout_imag = pymp.shared.array((self.number_of_nodes(), n_replica), dtype='float64')
with pymp.Parallel(self.num_threads) as p:
    for col in p.xrange(n_replica):
        x = _update_col(args)
        xout_real[:,col] = x.real.flatten()
        xout_imag[:,col] = x.imag.flatten()
xout = xout_real + 1j*xout_imag
```

**Comment by classner** [comment]
Totally makes sense. My comments were referring to having 'arbitrary complex' non-picklable datatypes for loop and iteration variables.

Your problem bugged me before and I found an elegant solution to it. The new code is tested and online on the master branch. I will not update the pypi version but wait and collect a few more features before increasing the version. In the meantime, you can safely use the master branch: I create `dev` branches for unstable development stages.

Let me know whether this solves your problem well, then I'll close the issue.

**Comment by classner** [comment]
Just a quick check: @kharris, is the issue resolved for you with the current version? :)

**Comment by kharris** [comment]
Yes, it works

**Comment by vasilipe1** [comment]
I am sorry to bother you. Could you please explain in more detail what needs to be done to work with shared arrays of complex type? The issue is closed, but I am not able to reproduce the solution. Many thanks.

**Comment by classner** [comment]
Hi @vasilipe1 ,

Can you share a minimal, executable example of what you're trying to do?

**Comment by vasilipe1** [comment]
I want to do the following (which does not work): 

MatEc = pymp.shared.array((Nq),dtype='complex')
with pymp.Parallel(num_cores) as p:
    for h in p.range(TDUnz):
          css = TDUred[h,0]
           iq =  TDUind[h,0]
          MatEc[iq] +=css

Instead of doing this (which works):

MatEc1 = pymp.shared.array((Nq),dtype='float')
MatEc2 = pymp.shared.array((Nq),dtype='float')
with pymp.Parallel(num_cores) as p:
    for h in p.range(TDUnz):
          css = TDUred[h,0]
           iq =  TDUind[h,0]
          MatEc1[iq] +=(css).real
          MatEc2[iq] +=(css).imag

Many thanks.

**Comment by classner** [comment]
Hi @vasilipe1 ,

Whereas the old change in principal made usage possible, there was a bug preventing it from working it  correctly. This is fixed with v0.4.3. Please confirm that it's working for you. :)

**Comment by vasilipe1** [comment]
Dear Christoph, 

Thank you very much! Yes, after reintalling pymp using  
pip install pymp-pypi  --upgrade --force-reinstall
my code with complex array 
MatEc = pymp.shared.array((Nq,Nq,Nw,Nw,Nw),dtype='complex') 
works great.

**Comment by classner** [comment]
Great to hear! Good luck with your project!

**Comment by vasilipe1** [comment]
Christoph, in case it is useful for anyone. 
In my example above, I have multiple nested 'for' loops. If I use 
for h in p.range(TDUnz):  
on the outer loop, I get the speedup, but the answers are different in parallel and single-core mode. 
When I use parallel only for the last inner 'for' loop, the answers are the same and the speedup is still there. 
(maybe a factor of 2 slower than in the ideal case.)
